scripts/js/lib/api/directMapXmlToMdx.ts

```diff
@@ -0,0 +1,127 @@
+// This code is a Qiskit project.
+//
+// (C) Copyright IBM 2025.
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+import type {
+  Root,
+  BlockContent,
+  Heading,
+  Text,
+  PhrasingContent,
+  List,
+  ListItem,
+  Paragraph,
+} from "mdast";
+
+// This type represents a node where all descendents have a 1:1 mapping with an mdx
+// node. For example:
+// * The XML '<title>' maps directly to MDX's `# title`. Therefore, a
+//   <title> node can be a descendent of the MdxMappableXmlNode.
+// * The XML `<parameterlist>` does NOT have a 1:1 mapping in MDX (we could map
+//   to a list, but how would we map back?). Therefore, MdxMappableXmlNodes can't
+//   have this as a descendent.
+type MdxMappableXmlNode =
+  | XmlTextNode
+  | XmlListNode
+  | { title: MdxMappableXmlNode[] }
+  | { para: MdxMappableXmlNode[] }
+  | { computeroutput: XmlTextNode[] }
+  | { verbatim: XmlTextNode[] };
+type XmlTextNode = { "#text": string };
+type XmlListNode = { itemizedlist: Array<{ listitem: MdxMappableXmlNode[] }> };
+
+/**
+ * Try to map the XML tree to a markdown AST (mdast) as closely as possible.
+ */
+export function directMapXmlToMdx(nodes: MdxMappableXmlNode[]): Root {
+  return {
+    type: "root",
+    children: nodes.flatMap((n) => xmlToBlockNodes(n, 1)),
+  };
+}
+
+/*
+ * Map an XML node (and it's children) to an array of mdast `Block` nodes
+ */
+function xmlToBlockNodes(
+  node: MdxMappableXmlNode,
+  headingDepth: Heading["depth"],
+): BlockContent[] {
+  if ("title" in node)
+    return [
+      {
+        type: "heading",
+        depth: headingDepth,
+        children: node.title.map(xmlToPhrasingNode),
+      },
+    ];
+  if ("para" in node) return xmlParagraphToBlockNode(node, headingDepth);
+  if ("verbatim" in node)
+    return [
+      {
+        type: "code",
+        lang: "C",
+        value: extractText(node.verbatim).trimEnd(),
+      },
+    ];
+  if ("itemizedlist" in node) return [xmlListToMdx(node)];
+  // If the node isn't a recognised block node, try wrapping in a paragraph and
+  // converting to a phrasing node.
```

Comment:
Would it be safer to error for now? I imagine we'll go through a big period of discovery with this script: new edge cases we haven't yet encountered. So defensive programming is highly valuable here

Reply:
Generally, yes, but I think in this case it's nice to be able to point the function at any XML and have it return a valid block node. Especially as the XML markup doesn't have the same concepts of block vs phrasing nodes.